add support for LTC7841

MCabuena opened the pull request:
## Pull Request Description

Please replace this with a detailed description and motivation of the changes. 
You can tick the checkboxes below with an 'x' between square brackets or just check them after publishing the PR. 
If this PR contains a breaking change, list dependent PRs and try to push all related PRs at the same time.

## PR Type
- [ ] Bug fix (change that fixes an issue)
- [x] New feature (change that adds new functionality)
- [ ] Breaking change (has dependencies in other repos or will cause CI to fail)

## PR Checklist
- [x] I have followed the [Coding style guidelines](http://analogdevicesinc.github.io/no-OS/drivers_guide.html#coding-style)
- [x] I have performed a self-review of the changes
- [x] I have commented my code, at least hard-to-understand parts
- [x] I have build all projects affected by the changes in this PR
- [x] I have tested in hardware affected projects, at the relevant boards
- [x] I have signed off all commits from this PR
- [x] I have updated the documentation (wiki pages, ReadMe etc), if applies

kister-jimenez commented:
/AzurePipelines run

MCabuena commented:
V2:
1. Removed from src.mk ltc7841_regs.c and ltc7841_regs.h requirement for building

kister-jimenez commented:
/AzurePipelines run

MCabuena commented:
V3:
1. fixed src.mk file inside ltc7841 project folder

CiprianRegus commented on drivers/power/ltc7841/ltc7841.h:
```diff
@@ -0,0 +1,199 @@
+/***************************************************************************//**
```
This file has the old license. Use the 3 clause BSD instead.

CiprianRegus commented on drivers/power/ltc7841/ltc7841.c:
```diff
@@ -0,0 +1,459 @@
+	uint8_t read_length;
+	uint8_t write_length = 1;
+	uint8_t size = get_register_size(cmd);
+	uint8_t write_buffer[write_length];
```
doesn't need to be an array

CiprianRegus commented on drivers/power/ltc7841/ltc7841.c:
```diff
@@ -0,0 +1,459 @@
+	ret = no_os_i2c_write(desc->comm_desc, write_buffer, write_length, 0);
+	if (ret)
+		return ret;
+	// read part of write_read, 1 at the end for stop bit sending
```
Use `/**/` comment style

CiprianRegus commented on drivers/power/ltc7841/ltc7841.c:
```diff
@@ -0,0 +1,459 @@
+	if (ret)
+		return ret;
+	// read part of write_read, 1 at the end for stop bit sending
+	ret = no_os_i2c_read(desc->comm_desc, g_i2c0_read_buffer, read_length, 1);
```
You should store the register value in `data`.

CiprianRegus commented on drivers/power/ltc7841/ltc7841.c:
```diff
@@ -0,0 +1,459 @@
+{
+	int ret;
+	/* Write register */
+	ret = ltc7841_reg_write(desc, LTC7841_OPERATION, LTC7841_OPERATION_MARGIN_LOW);
```
Return directly. There are other places where this is the case.

kister-jimenez commented:
/AzurePipelines run

CiprianRegus commented on drivers/power/ltc7841/iio_ltc7841.c:
```diff
@@ -0,0 +1,428 @@
+	return iio_format_value(buf, len, IIO_VAL_INT_PLUS_MICRO, 1, vals);
+}
+
+static struct iio_attribute ltc7841_chan_attrs[] = {
+	{
+		.name = "raw",
+		.show = ltc7841_iio_read_raw
+	},
+	END_ATTRIBUTES_ARRAY
+};
```
Is the scale of every channel equal to 1? Otherwise, you also need the scale attribute here.

MCabuena replied:
The scale of every channel is the same at 10mV per bit.

MCabuena replied:
on my next commit, I'll add as you stated below a scale attribute for the channel

CiprianRegus replied:
When implementing review suggestions, you don't need to create a new commit (unless you add new functionality, which should have its own commit). Just modify the current one and push the changes again. After that, you should add a changelog comment in this PR stating what you did since the last push.

CiprianRegus commented on drivers/power/ltc7841/iio_ltc7841.c:
```diff
@@ -0,0 +1,428 @@
+	struct ltc7841_iio_desc *iio_ltc7841 = dev;
+	struct ltc7841_desc *ltc7841 = iio_ltc7841->ltc7841_desc;
+
+	ret = ltc7841_reg_read(ltc7841, (uint8_t)reg, (uint8_t *)readval);
```
Use an intermediary `uin8_t` variable. Currently, you're passing a different memory location based on endianess.

CiprianRegus commented on drivers/power/ltc7841/iio_ltc7841.c:
```diff
@@ -0,0 +1,428 @@
+	}
+	case LTC7841_IIO_IIN_CHAN:
+		ret = ltc7841_reg_read(ltc7841, (uint8_t)LTC7841_READ_IIN, read_value);
+		if (0 == ret) {
```
Use inverse logic:
```
if (ret)
    break;
```

CiprianRegus commented on drivers/power/ltc7841/iio_ltc7841.c:
```diff
@@ -0,0 +1,428 @@
+	case LTC7841_IIO_IIN_CHAN:
+		ret = ltc7841_reg_read(ltc7841, (uint8_t)LTC7841_READ_IIN, read_value);
+		if (0 == ret) {
+			double current_value;
```
Use a lower unit (e.g microampere in this case), so you won't lose any digit during the division, and scale the result instead.

MCabuena commented:
V4:
1. changed commenting style to /* */
2. removed void in common_data.c for ltc7841_i2c_extra
3. removed PEC_SIZE since this device doesn't support PEC
4. added LTC7841_ name to BYTE_LENGTH and WORD_LENGTH in ltc7841.h
5. when computing for current input and current output, unit usage is now in uA for initial calculations, only when calculating for scale values does it show in A
6. changed iio_ltc7841.c all reverse logic for checking ret
7. added scaled value as an attribute
8. functions in ltc7841.c now return directly
9. changed write_buffer into a uint8_t type only, and not as an array
10. changed license file of ltc7841.h

buha commented:
/AzurePipelines run

buha commented on drivers/power/ltc7841/ltc7841.c:
```diff
@@ -0,0 +1,435 @@
+	}
+	return ret;
+}
+/**
```
it's a style remark but always leave a blank line from the closing bracket of a function to the next content like this
```
}

/**
```

reason is, we want the code style to be consistent

buha commented on drivers/power/ltc7841/ltc7841.c:
```diff
@@ -0,0 +1,435 @@
+	} else {
+		ret = -EINVAL;
+	}
+	return ret;
```
another style remark, it's kind of informal but we always put a blank line before the final return statement of the function like:
```
    }

    return ret;
}
```

buha commented on drivers/power/ltc7841/ltc7841.h:
```diff
@@ -0,0 +1,191 @@
+ */
+struct ltc7841_init_param {
+	/* Device register settings */
+	struct ltc7841_reg_st *regs;
```
i cant find where this struct comes from

buha replied:
is the user supposed to provide the register configuration explicitly ? i dont think this approach is best, i think the driver should abstract away as much as possible the registers and not force the user to read the datasheet and provide register values

MCabuena replied:
Hi buha,

It used to be that I have a list of registers I put on a list and populated that. I changed the approach and abstracted it. I wasn't able to remove it on this version of the commit.

I'll remove that part for the next version which I plan to also address your comments.

buha commented on projects/ltc7841/src.mk:
```diff
@@ -0,0 +1,46 @@
+INCS += $(PROJECT)/src/platform/$(PLATFORM)/parameters.h
+SRCS += $(PROJECT)/src/platform/$(PLATFORM)/parameters.c
+
+INCS += $(INCLUDE)/no_os_delay.h     		\
```
use NO_OS_INC_DIRS instead of INCS, search no-OS github to see examples of its usage

buha commented on projects/ltc7841/src/platform/maxim/platform_src.mk:
```diff
@@ -0,0 +1,16 @@
+INCS += $(PLATFORM_DRIVERS)/maxim_gpio.h      \
```
same here, NO_OS_INC_DIRS

MCabuena commented:
V5:
1. changed src.mk file to use NO_OS_INC_DIRS instead of INC +=
2. removed ltc7841_reg_st inside ltc7841_init_param
3. added a blank space before the final return as suggested
4. added a space in between the {} of functions

kister-jimenez commented:
/AzurePipelines run

buha commented on projects/ltc7841/src/common/common_data.h:
```diff
@@ -0,0 +1,52 @@
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *  - Redistributions of source code must retain the above copyright
```
you still have some of these old license headers, please check them all

buha commented on projects/ltc7841/src/examples/basic/basic_example.h:
```diff
@@ -0,0 +1,44 @@
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *  - Redistributions of source code must retain the above copyright
```
this one

buha commented on projects/ltc7841/src/platform/maxim/parameters.c:
```diff
@@ -0,0 +1,54 @@
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *  - Redistributions of source code must retain the above copyright
```
this one, and potentially others

buha commented on projects/ltc7841/src/platform/platform_includes.h:
```diff
@@ -0,0 +1,46 @@
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *  - Redistributions of source code must retain the above copyright
```
this one

buha commented on projects/ltc7841/src/platform/maxim/platform_src.mk:
```diff
@@ -0,0 +1,16 @@
+        $(PLATFORM_DRIVERS)/../common/maxim_dma.h       \
+        $(PLATFORM_DRIVERS)/maxim_i2c.h       \
+        $(PLATFORM_DRIVERS)/maxim_uart.h      \
+        $(PLATFORM_DRIVERS)/maxim_uart_stdio.h
```
you still don't use NO_OS_INC_DIRS here

buha commented on projects/ltc7841/README.rst:
```diff
@@ -0,0 +1,128 @@
+.. code-block:: bash
+
+	# Select the example you want to enable by choosing y for enabling and n for disabling
+	BASIC_EXAMPLE = y
```
i don't think editing Makefiles is good to set the example to build even though we have existing projects done this way... let's strive to make the user interaction easy so that they run one of these commands to choose the example to build:

```
make EXAMPLE=basic_example
make EXAMPLE=iio_example
```

Basically you could default to basic like this:
```
EXAMPLE ?= basic_example
```

And take action based on it:
```
ifeq ($(INTERFACE), usb)
...
endif
```
```

kister-jimenez commented on drivers/power/ltc7841/iio_ltc7841.c:
```diff
@@ -0,0 +1,493 @@
+	int32_t value;
+
+	switch (channel->address) {
+	case LTC7841_IIO_VIN_CHAN: {
+		ret = ltc7841_reg_read(ltc7841, (uint8_t)LTC7841_READ_VIN, read_value);
+		if (ret)
+			break;
+		temp = no_os_get_unaligned_le16(read_value);
+		break;
+	}
+	case LTC7841_IIO_VOUT_CHAN: {
+		ret = ltc7841_reg_read(ltc7841, (uint8_t)LTC7841_READ_VOUT, read_value);
+		if (ret)
+			break;
+		temp = no_os_get_unaligned_le16(read_value);
+		break;
+	}
+	case LTC7841_IIO_IIN_CHAN:
+		ret = ltc7841_reg_read(ltc7841, (uint8_t)LTC7841_READ_IIN, read_value);
+		if (ret)
+			break;
+		temp = no_os_get_unaligned_le16(read_value);
+		break;
+	case LTC7841_IIO_IOUT_CHAN:
+		ret = ltc7841_reg_read(ltc7841, (uint8_t)LTC7841_READ_IOUT, read_value);
+		if (ret)
+			break;
+		temp = no_os_get_unaligned_le16(read_value);
+		break;
+	case LTC7841_IIO_TEMP_CHAN:
+		ret = ltc7841_reg_read(ltc7841, (uint8_t)LTC7841_READ_TEMPERATURE_1,
+				       read_value);
+		if (ret)
+			break;
+		temp = no_os_get_unaligned_le16(read_value);
+		break;
+	default:
+		return -EINVAL;
```
You might want to be consistent with your brackets in case statements. But, I don't see a need to add one.

kister-jimenez replied:
brackets use in the case statements are still in consistent.

kister-jimenez commented on drivers/power/ltc7841/iio_ltc7841.c:
```diff
@@ -0,0 +1,493 @@
+	uint16_t temp;
+
+	switch (channel->address) {
+	case LTC7841_IIO_VIN_CHAN: {
+		ret = ltc7841_reg_read(ltc7841, (uint8_t)LTC7841_READ_VIN, read_value);
+		if (ret)
+			break;
+		temp = no_os_get_unaligned_le16(read_value);
+		value[0] = (int32_t)(temp / CENTI);
+		value[1] = (int32_t)(temp % CENTI) * DECIMAL_POINT_SCALE_VOLTAGE;
+		ret = iio_format_value(buf, len, IIO_VAL_INT_PLUS_MICRO, 2, value);
+		break;
+	}
+	case LTC7841_IIO_VOUT_CHAN: {
+		ret = ltc7841_reg_read(ltc7841, (uint8_t)LTC7841_READ_VOUT, read_value);
+		if (ret)
+			break;
+		temp = no_os_get_unaligned_le16(read_value);
+		value[0] = (int32_t)(temp / CENTI);
+		value[1] = (int32_t)(temp % CENTI) * DECIMAL_POINT_SCALE_VOLTAGE;
+		ret = iio_format_value(buf, len, IIO_VAL_INT_PLUS_MICRO, 2, value);
+		break;
+	}
```
same as previous comment.

kister-jimenez commented on drivers/power/ltc7841/ltc7841.h:
```diff
@@ -0,0 +1,189 @@
+#define LTC7841_WORD_LENGTH         2
+#define LTC7841_I2C_WR_FRAME_SIZE   3
+/* this denotes the rsense scale of LTC7841 2(both channels) * 50 uV/ bit / 0.002 (rsense resistor), unit in uA*/
+#define CURRENT_SCALE               (20)
```
Since RSENSE will now be a param value, you can compute this instead.

kister-jimenez replied:
Also it's not used in the driver but in the iio_driver, maybe just add this as the scale attribute in the iio_driver.

kister-jimenez commented on drivers/power/ltc7841/ltc7841.h:
```diff
@@ -0,0 +1,189 @@
+#define WRITE_ONLY_REGISTERS_NUMBER 2
+#define ECOMM                       9
+/* specific only to DC2798A in ohms*/
+#define RSENSE_VALUE_IIN            (500)
```
This might be something you will need to add as param in the example application. Anything that is specific to the board level (i.e. outside the chip) add it as param. However, this is not used anywhere in you driver and iio_driver.

kister-jimenez commented on drivers/power/ltc7841/iio_ltc7841.c:
```diff
@@ -0,0 +1,493 @@
+	/* final percentage value to be written to register, scaling done on the last part*/
+	vout_value = (uint16_t)(temp_value /
+				CENTI);
+	return ltc7841_change_vout_command(ltc7841, vout_value);
```
Add extra newline before the return for readability

kister-jimenez commented on drivers/power/ltc7841/iio_ltc7841.c:
```diff
@@ -0,0 +1,493 @@
+{
+	struct ltc7841_iio_desc *iio_ltc7841 = dev;
+	struct ltc7841_desc *ltc7841 = iio_ltc7841->ltc7841_desc;
+	return ltc7841_clear_status_word_bits(ltc7841);
```
Add extra newline before the return for readability.

kister-jimenez commented on drivers/power/ltc7841/iio_ltc7841.c:
```diff
@@ -0,0 +1,493 @@
+	ret = ltc7841_reg_read(ltc7841, (uint8_t)LTC7841_READ_VOUT, read_value);
+	if (ret)
+		return ret;
+	temp = no_os_get_unaligned_le16(read_value);
```
You may want to add a new line after the return for better readability.

kister-jimenez commented on drivers/power/ltc7841/iio_ltc7841.c:
```diff
@@ -0,0 +1,493 @@
+	.debug_attributes = ltc7841_debug_attrs,
+	.debug_reg_read = ltc7841_iio_reg_read,
+	.debug_reg_write = ltc7841_iio_reg_write,
+};
```
And extra newline at EOF.

kister-jimenez commented on projects/ltc7841/src.mk:
```diff
@@ -0,0 +1,26 @@
+		$(NO-OS)/util/no_os_mutex.c	\
+		$(NO-OS)/util/no_os_crc8.c
+
+SRCS += $(DRIVERS)/power/ltc7841/ltc7841.c
```
Add extra newline

kister-jimenez commented on projects/ltc7841/src/common/common_data.c:
```diff
@@ -0,0 +1,64 @@
+
+struct ltc7841_init_param ltc7841_ip = {
+	.comm_param = &ltc7841_i2c_ip
+};
```
Add extra newline

kister-jimenez commented on projects/ltc7841/src/examples/iio_example/iio_example.c:
```diff
@@ -0,0 +1,80 @@
+	if (ret)
+		pr_err("Error!\n");
+	return ret;
+}
```
Add extra newline

kister-jimenez commented on drivers/power/ltc7841/ltc7841.h:
```diff
@@ -0,0 +1,189 @@
+extern const uint8_t ltc7841_write_only_registers[WRITE_ONLY_REGISTERS_NUMBER];
+
+/* LTC7841 registers list */
+enum ltc7841_registers {
+	LTC7841_OPERATION = 1,
+	LTC7841_VOUT_MODE = 0x20,
+	LTC7841_STATUS_WORD = 0x79,
+	LTC7841_READ_VIN = 0x88,
+	LTC7841_READ_IIN = 0x89,
+	LTC7841_READ_VOUT = 0x8B,
+	LTC7841_READ_IOUT = 0x8C,
+	LTC7841_READ_TEMPERATURE_1 = 0x8D,
+	LTC7841_PMBUS_REVISION = 0x98,
+	LTC7841_MFR_IOUT_PEAK = 0xD7,
+	LTC7841_MFR_VOUT_PEAK = 0xDD,
+	LTC7841_MFR_VIN_PEAK = 0xDE,
+	LTC7841_MFR_TEMEPRATURE1_PEAK = 0xDF,
+	LTC7841_MFR_IIN_PEAK = 0xE1,
+	LTC7841_MFR_CLEAR_PEAKS = 0xE3,
+	LTC7841_MFR_VOUT_MARGIN_HIGH = 0xE5,
+	LTC7841_MFR_SPECIAL_ID = 0xE7,
+	LTC7841_MFR_VOUT_COMMAND = 0xE8,
+	LTC7841_MFR_CONFIG = 0xE9,
+	LTC7841_MFR_VOUT_MARGIN_LOW = 0xED,
+	LTC7841_MFR_RAIL_ADDRESS = 0xFA,
+	LTC7841_MFR_RESET = 0xFD
+};
```
You may want to follow this: https://github.com/analogdevicesinc/no-OS/blob/a75c351f56ab2699d7b0fe8623ffcaf4e867f6ec/drivers/power/lt7182s/lt7182s.h#L50C1-L170C33

For consistency across PMBUS devices and other no-OS drivers.

MCabuena commented:
I wanted to re-open this pull request

MCabuena commented:
V6:
1. changed enum registers for pmbus into # define
2. changed, checked and re-checked files to use BSD 3 clause for license files
3. adjusted driver README.rst to reflect the current ltc7841.h file
4. adjusted project README.rst to instruct using new command for building the examples, adding a variable of EXAMPLE=IIO_EXAMPLE or EXAMPLE=BASIC_EXAMPLE
5. changed examples_src.mk to use NO_OS_INC_DIRS
6. changed examples_src.mk to check if EXAMPLE variable is equal to IIO_EXAMPLE or EXAMPLE is equal to BASIC_EXAMPLE
7. added blank line in files as the last line
8. added space before final return

kister-jimenez commented:
/AzurePipelines run

MCabuena commented:
V7:
1. changed builds.json to add new EXAMPLE variable, EXAMPLE=BASIC_EXAMPLE for basic example and EXAMPLE=IIO_EXAMPLE for iio example

kister-jimenez commented:
/AzurePipelines run

kister-jimenez commented on projects/ltc7841/src/examples/examples_src.mk:
```diff
@@ -0,0 +1,24 @@
+ifeq ($(EXAMPLE), IIO_EXAMPLE))
```
remove extra ')'

kister-jimenez commented on projects/ltc7841/src/examples/examples_src.mk:
```diff
@@ -0,0 +1,24 @@
+SRCS += $(PROJECT)/src/examples/iio_example/iio_example.c
+endif
+
+ifeq ($(EXAMPLE), BASIC_EXAMPLE))
```
remove extra ')'

kister-jimenez commented on projects/ltc7841/Makefile:
```diff
@@ -0,0 +1,9 @@
+# Select the example you want to enable by choosing y for enabling and n for disabling
+BASIC_EXAMPLE = n
+IIO_EXAMPLE = y
```
You need to remove this and replace with maybe a default like this:
EXAMPLE ?= IIO_EXAMPLE

kister-jimenez commented on projects/ltc7841/src/platform/maxim/main.c:
```diff
@@ -0,0 +1,72 @@
+#endif
+
+#if (BASIC_EXAMPLE + IIO_EXAMPLE == 0)
+#error At least one example has to be selected using y value in Makefile.
```
Need to update error messages because EXAMPLE is now passed as an argument to make instead of editing the makefile.
e.g.

Select an example by passing the type of EXAMPLE in the make command. For example:
make EXAMPLE=IIO_EXAMPLE

kister-jimenez commented on drivers/power/ltc7841/iio_ltc7841.c:
```diff
@@ -0,0 +1,493 @@
+ * @return ret    - Result of the reading procedure.
+ * 		    In case of success, the size of the read data is returned.
+*/
+//static
```
//static is still present

kister-jimenez commented on drivers/power/ltc7841/ltc7841.c:
```diff
@@ -0,0 +1,465 @@
+{
+	struct ltc7841_desc *dev;
+	int ret;
+	/* ltc7841 only needs sda, scl and ground connection to the microcontroller*/
```
I don't think this comment is necessary. You can remove this.

MCabuena replied:
comment removed

kister-jimenez commented on drivers/power/ltc7841/ltc7841.c:
```diff
@@ -0,0 +1,465 @@
+	/* I2C write target address */
+	read_length = size;
+	write_buffer = cmd;
+	/* write part of write_read, 0 at the end for no stop bit */
```
You can remove this comment because these can be inferred from line.

MCabuena replied:
comment removed

kister-jimenez commented on drivers/power/ltc7841/ltc7841.c:
```diff
@@ -0,0 +1,465 @@
+	ret = no_os_i2c_write(desc->comm_desc, &write_buffer, 1, 0);
+	if (ret)
+		return ret;
+	/* read part of write_read, 1 at the end for stop bit sending */
```
You can remove this comment as well.

MCabuena replied:
comment removed

kister-jimenez commented on drivers/power/ltc7841/ltc7841.c:
```diff
@@ -0,0 +1,465 @@
+	// apply a 7 bit mask since bit 7 needs to be held low to enable communication
+	addr &= LTC7841_7_BIT_ADDRESS_MASK;
+
+	/* Write register */
```
you can remove this comment.

MCabuena replied:
comment removed

kister-jimenez commented on drivers/power/ltc7841/ltc7841.c:
```diff
@@ -0,0 +1,465 @@
+ */
+int ltc7841_change_rail_address(struct ltc7841_desc *desc, uint8_t addr)
+{
+	// apply a 7 bit mask since bit 7 needs to be held low to enable communication
```
you can remove this comment.

MCabuena replied:
comment removed

kister-jimenez reviewed (commented):
There are still previous comments that have not been addressed in this revision. Please address and resolve them all.

buha commented on projects/ltc7841/Makefile:
```diff
@@ -0,0 +1,9 @@
+# Select the example you want to enable by choosing y for enabling and n for disabling
+BASIC_EXAMPLE = n
```
Please see this pull request and use it as a template: https://github.com/analogdevicesinc/no-OS/pull/2368

The idea is that the user types `EXAMPLE=basic` or `EXAMPLE=iio_example`, these being the names of the folders where you put the example code. We will create some infrastructure to do this automatically, but for now you'll have to do it manually.

kister-jimenez commented:
/AzurePipelines run

kister-jimenez commented:
/AzurePipelines run

MCabuena commented:
V8
1. In switch case statements, I removed the {} inside the cases.
2. I followed buha to use 2368 as a template for adding EXAMPLE as a variable
3. changes to makefile and examples_src.mk to use EXAMPLE

kister-jimenez commented:
/AzurePipelines run

MCabuena commented:
Hi,

I'm not sure why my README file inside drivers/power/README.rst cannot be found. I'll try recommitting the same files and check if the issue happens again.

That's the cause of the issue, but if you check the 3rd commit, the file is included.

kister-jimenez commented:
> Hi,
> 
> I'm not sure why my README file inside drivers/power/README.rst cannot be found. I'll try recommitting the same files and check if the issue happens again.
> 
> That's the cause of the issue, but if you check the 3rd commit, the file is included.

You have a typo in part number inside the README.rst.

kister-jimenez commented:
/AzurePipelines run

kister-jimenez commented on projects/ltc7841/src/examples/iio_example/iio_example.c:
```diff
@@ -0,0 +1,80 @@
+
+	ret = ltc7841_iio_init(&ltc7841_iio_desc, &ltc7841_iio_ip);
+	if (ret)
+		goto exit;
+	struct iio_app_device iio_devices[] = {
```
nit: add newline.

MCabuena commented:
V9
1. changed CURRENT_SCALE to manual computation with a parameter input for rsense : RSENSE_VALUE_DC2798A, in case user wants to use a different rsense value from DC2798A, he can change the input to common_data.c for the LTC7841_rsense_value
2. removed astyle.exe files and astyle_config files
3. changed int32_t type to int32 in iio_ltc7841.c file for the return type
4. hopefully all nits?

kister-jimenez commented:
/AzurePipelines run

buha reviewed (changes requested):
It looks good but you'll have to update the project to the new project infrastructure.

Here's the PR with the changes with 3 modified projects as example: https://github.com/analogdevicesinc/no-OS/pull/2412

And here's a pending PR where I gave feedback with pretty much that needs to be changed, same applies to your PR.

kister-jimenez commented on projects/ltc7841/Makefile:
```diff
@@ -0,0 +1,5 @@
+include ../../tools/scripts/generic_variables.mk
+
```
Add:

```
EXAMPLE := iio_example
include ../../tools/scripts/examples.mk
```

These lines sets the default EXAMPLE in case not provided and includes the source/include files for the examples.

kister-jimenez commented on projects/ltc7841/src/platform/platform_includes.h:
Delete this and change all all occurences of #include "platform_includes.h" in your project files to #include "parameters.h"

kister-jimenez commented on projects/ltc7841/src.mk:
```diff
@@ -0,0 +1,26 @@
+include $(PROJECT)/src/platform/$(PLATFORM)/platform_src.mk
+include $(PROJECT)/src/examples/examples_src.mk
+
+SRCS += $(PROJECT)/src/platform/$(PLATFORM)/main.c
+SRCS += $(PROJECT)/src/common/common_data.c
+SRCS += $(PROJECT)/src/platform/$(PLATFORM)/parameters.c
```
You can delete this as this is now handled by the example infra script.

kister-jimenez commented on projects/ltc7841/src/examples/examples_src.mk:
You can delete this now and add mk files in each example folder.
for example in the iio_example, you can maybe add an iio_example.mk. Name won't matter as long as file extension is .mk.

Then inside it:
```
IIOD=y

INCS += $(DRIVERS)/power/ltc7841/iio_ltc7841.h
SRCS += $(DRIVERS)/power/ltc7841/iio_ltc7841.c	\	
```

iio related files are handled also by the infra scripts.

basic example don't need this mk file unless you need example specific defines and sources/includes.

kister-jimenez commented on projects/ltc7841/src/platform/maxim/main.c:
```diff
@@ -0,0 +1,65 @@
+	int ret = -EINVAL;
+
+#ifdef BASIC_EXAMPLE
+	struct no_os_uart_desc *uart_desc;
+
+	ret = no_os_uart_init(&uart_desc, &uart_ip);
+	if (ret)
+		return ret;
+
+	no_os_uart_stdio(uart_desc);
```
Move this to the basic_example.c `example_main()`

kister-jimenez commented on projects/ltc7841/src/platform/maxim/main.c:
```diff
@@ -0,0 +1,65 @@
+{
+	int ret = -EINVAL;
+
+#ifdef BASIC_EXAMPLE
+	struct no_os_uart_desc *uart_desc;
+
+	ret = no_os_uart_init(&uart_desc, &uart_ip);
+	if (ret)
+		return ret;
+
+	no_os_uart_stdio(uart_desc);
+	ret = basic_example_main();
+#endif
+
+#ifdef IIO_EXAMPLE
+	ret = iio_example_main();
+#endif
```
You can replace this now with just the common `example_main()` function

kister-jimenez reviewed (commented):
The EXAMPLE=<example> has been implemented in the previous comments so I just added the comments related to the other changes needed for the recently merged examples infra.

Addition minor changes on the rsense placement.

MCabuena commented:
V10
1. moved rsense parameter to ltc7841_iio_init parameter
2. followed Kister's instructions to change the project files

kister-jimenez commented:
/AzurePipelines run

MCabuena commented:
V11
1. removed \ at the end of SRCS += $(DRIVERS)/power/ltc7841/iio_ltc7841.c in examples_src.mk

kister-jimenez commented:
/AzurePipelines run

MCabuena commented:
V12
1. performed astyle.exe --options= on basic_example.c

kister-jimenez commented:
/AzurePipelines run

MCabuena commented on drivers/power/ltc7841/iio_ltc7841.c:
```diff
@@ -0,0 +1,501 @@
+	if (!desc)
+		return -ENODEV;
+
+	no_os_free(desc->iio_dev->channels);
```
I should just remove this no_os_free? I thought this was always dynamically allocated. When did this change occur?

CiprianRegus replied:
You're assigning this pointer here: https://github.com/analogdevicesinc/no-OS/blob/4d0014652cc29dde91b0e42227c7f6d502bb7d1a/drivers/power/ltc7841/iio_ltc7841.c#L495 . It was never dynamically allocated.

MCabuena commented:
V13
1. removed no_os_free for channels

buha commented:
/AzurePipelines run

buha commented on drivers/power/ltc7841/ltc7841.c:
```diff
@@ -0,0 +1,461 @@
+	ret = no_os_i2c_init(&dev->comm_desc, init_param->comm_param);
+	if (ret)
+		goto i2c_err;
+	/*initialize mfr vout command to half upon initializing*/
```
try to format comments like:
```
   capital
   |
/* Text */
  ^    ^
   whitespace
```

reasoning being to have consistent looking comments, otherwise we end up with as many comment styles as contributors

buha commented on drivers/power/ltc7841/ltc7841.c:
```diff
@@ -0,0 +1,461 @@
+ */
+int ltc7841_reset(struct ltc7841_desc *desc)
+{
+
```
remove blank line

buha commented on drivers/power/ltc7841/ltc7841.c:
```diff
@@ -0,0 +1,461 @@
+	if (vout_level <= MFR_MAX_VALUE) {
+		/* okay to proceed */
+		ret = ltc7841_reg_write(desc, LTC7841_MFR_VOUT_MARGIN_HIGH, vout_level);
+		if (ret)
```
delete line 426 and 427, it comes down to the same result

buha commented on drivers/power/ltc7841/iio_ltc7841.c:
```diff
@@ -0,0 +1,500 @@
+	struct ltc7841_desc *ltc7841 = iio_ltc7841->ltc7841_desc;
+
+	ret = ltc7841_reg_read(ltc7841, (uint8_t)reg, register_value);
+	*readval = no_os_get_unaligned_le16(register_value);
```
i'm not sure what's happening here, readval is a `uint32_t *`, i'm pretty sure on one of the endianesses this will fail

buha commented on projects/ltc7841/src/platform/maxim/parameters.h:
```diff
@@ -0,0 +1,55 @@
+
+/* LTC7841 parameters */
+#define LTC7841_I2C_DEVICE_ID                  (0)
+#define LTC7841_I2C_CLK_SPEED                  (400000)
```
parenthesize only when it's an expression with macro arguments, not when it's a constant

MCabuena commented:
V14
1. removed ret variable in main.c
2. removed Include Files comment
3. changed formatting of comments
4. removed parenthesis in #define for constants
5. added intermediary value for uint16_t to be used for alignment before passing the value to uint32_t
6. removed else statement for returning invalid condition

kister-jimenez commented:
/AzurePipelines run

MCabuena commented:
V15
1. Added ret = -EINVAL for initialization
2. fix astyle of iio_ltc7841.c

kister-jimenez commented:
/AzurePipelines run

MCabuena commented:
V16
1. removed blank space before return for 1 liner functions

kister-jimenez commented:
/AzurePipelines run

kister-jimenez commented on projects/ltc7841/src/examples/iio_example/iio_example.mk:
```diff
@@ -0,0 +1,4 @@
+IIOD=y
+
+INCS += $(DRIVERS)/power/ltc7841/iio_ltc7841.h
+SRCS += $(DRIVERS)/power/ltc7841/iio_ltc7841.c
```
NIT: Add newline at the end

kister-jimenez commented on projects/ltc7841/src/examples/basic/basic_example.c:
```diff
@@ -0,0 +1,125 @@
+/***************************************************************************//**
+ *   @file   basic_example.c
+ *   @brief  Basic example source file for ltp8800 project.
```
ltp8800-->ltc7841

kister-jimenez reviewed (changes requested):
Some minor changes in my comments. You might want to update the copyright on all the files to 2025.

Other than that, it looks good to me.

amiclaus commented on projects/ltc7841/src/examples/basic/basic_example.c:
```diff
@@ -0,0 +1,125 @@
+	while (1) {
+		error = ltc7841_mfr_clear_peaks(ltc7841_desc);
+		if (error)
+			return error;
```
these should be conditional jumps outside the while loop to free resources.

MCabuena commented on drivers/power/ltc7841/ltc7841.c:
```diff
@@ -0,0 +1,449 @@
+{
+	int ret;
+	uint8_t read_length;
+	uint8_t size = get_register_size(cmd);
```
This is okay, this will fail in no_os_i2c_read

MCabuena replied:
Added a check that size should not be equal to 0 for read to continue, otherwise -EINVAL will be returned

MCabuena commented:
V17
1. changed copyright year to 2025
2. changed ltp8800 to ltc7841 in comments
3. added blank space at the end of file for iio_example.mk
4. added check for size in reg read
5. removed redundant return ret
6. added remove_ltc7841 as conditional jump

jemfgeronimo commented:
/AzurePipelines run

kister-jimenez commented:
Please rebase to latest no-os main.

CiprianRegus commented:
/AzurePipelines run

kister-jimenez commented:
/AzurePipelines run

MCabuena commented:
hi, I would like to request if we can merge this
